### ISSUE TITLE
Filter fastq does not handle file names correctly

### ISSUE THREAD

**Issue by mbhall88** [esteinig/sketchy#20]
https://github.com/esteinig/sketchy/blob/66a2004223bcd72481f5cda875d39341f15b3f2a/sketchy/utils.py#L112-L131

`filter_fastq` does not fail if the input file is not a valid fastq format.

**Comment by mbhall88** [comment]
Actually, on further thought, failing is probably not the best behaviour in such a case. Outputting nothing (as you are currently doing) is probably the best choice.

**Comment by mbhall88** [comment]
In addition, the current method to check if the record name exists is inefficient. In the worst case, we search the entire list of required record names once for each record in the fastq file meaning in the worst case it is `O(n * m)` where `n` is the length of `records` and `m` is the number of records in the fastq file. This is because `records` is a `list`. It would be much more efficient to use a `set` and this would make the runtime `O(m)`.

**Comment by mbhall88** [comment]
I have a fix for this issue locally and will push once #19 is merged.

**Comment by esteinig** [comment]
Thanks for those fixes Michael, super appreciated. 

This task was more like a support function for now and will get proper attention for the mixture analysis mode.

Thanks for the `list` vs `set` information on runtime, I should have considered that for sure!

Regarding name strip, are you sure it's not parsing the @ in the `sketchy.utils.query_fastq` function which provides the records for `filter_fastq`?

I remember vaguely keeping it in, my apologies, on mobile and on flight.

**Comment by esteinig** [comment]
Can you explain briefly why it's better to fail silently than raising an error on wrong input format?

**Comment by mbhall88** [comment]
Ah, well if you're expecting to have the `@` symbol in then it would work. But that seems slightly unconventional?  
The changes I have made switch this to using `pysam` which has the added benefit of allowing seamless handling of fasta too.

> Can you explain briefly why it's better to fail silently than raising an error on wrong input format?

Most libraries I have worked with, such as `pysam` do this. Although I guess that doesn't mean we should accept it too. If you think we should fail on incorrect formatting then we can do that?
